### ISSUE TITLE
Online topological sort

### ISSUE THREAD

**Issue by adriangb** [Qiskit/retworkx#495]
### What is the expected enhancement?
I'm interested in replacing [graphlib.TopologicalSorter](https://docs.python.org/3/library/graphlib.html) with something that will also let me do other graph operations (namely pruning subgraphs based on nodes, which it seems like I can do with `PyDAG.substitute_node_with_subgraph`).
In particular, this sort of topological sort allows you to poll for available dependencies as you process them, vs. just giving you a topologically sorted list of all of the nodes in the graph.

**Comment by georgios-ts** [comment]
Thanks for opening this. Can you elaborate a bit more on the usecase you have in mind?
Adding a functionality like `graphlib.TopologicalSorter` should be possible but one key limitation might be that you will *not* be able to add/remove nodes/modify the graph while you iterate over it. For reference, we recently added an "online" BFS traversal https://github.com/Qiskit/retworkx/blob/main/retworkx/__init__.py#L1739-L1808 that it'll be in the next release.

**Comment by adriangb** [comment]
Hi, thank you for the quick reply!

> Can you elaborate a bit more on the usecase you have in mind?

My use case is for a dependency injection library. Dependencies could take ns to execute or ms, so simply following a predefined topological order is inefficient. I basically need to make dependencies available for processing as they are freed up (freed up meaning they have no children / predecessors). FWIW my current implementation of this is [here](https://github.com/adriangb/di/blob/60445225ba9945a9d657864ebb776e4c8cfa1a8e/di/_utils/task.py#L94-L108). The reason I don't want to use `ToplogicalSorter` itself is that I need to make copies of the "solved" (checked for cycles, etc.) state so that I can execute it in a tight loop. I also eventually want to be able to do Graph operations on it.

> one key limitation might be that you will not be able to add/remove nodes/modify the graph while you iterate over it

Logically! In fact that is how `TopologicalSorter` works: you can't iterate over it until you call `prepare()`, and once you call `prepare()` you can't add any more nodes. It does not have the ability to remove nodes.

> we recently added an "online" BFS traversal

Yes, I saw that while I was searching for this issue. The callback / visitor API looks nice (great work btw), although I'm not sure if it'll work out for an online topological sort as described here.

**Comment by adriangb** [comment]
This is great!
In between the time I made this issue and now I wrote my own implementation: https://github.com/adriangb/graphlib2
The main differences vs. this implementation is 1:1 API parity with the stdlib, including supporting Python objects as nodes in the graph.

Now that this got merged, I added benchmarks comparing this vs. the stdlib vs. my implementation: https://github.com/adriangb/graphlib2/blob/main/bench.ipynb

Looks like this implementation performs considerably better than the stdlib, and about the same as my implementation. Nice work (to both of us)!
